### PR TITLE
[Core] Move default taxation settings to channel

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: features/channel/managing_channels/selecting_default_tax_calculation_strategy_for_channel.feature [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    As an Administrator
+    I want to be able to select default tax calculation strategy
+
+    Background:
+        Given I am logged in as an administrator
+
+    @ui
+    Scenario: Adding a new channel with implicitly selected default tax calculation strategy
+        Given I want to create a new channel
+        When I specify its code as MOBILE
```
Comment (pamil) [inline features/channel/managing_channels/selecting_default_tax_calculation_strategy_for_channel.feature]: "MOBILE"?

File: features/channel/managing_channels/selecting_default_tax_calculation_strategy_for_channel.feature [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    @ui
+    Scenario: Adding a new channel with default tax calculation strategy
+        Given I want to create a new channel
+        When I specify its code as MOBILE
+        And I select the "Order item units based" as default tax calculation strategy
+        And I name it "Mobile store"
+        And I add it
+        Then I should be notified that it has been successfully created
+        And "Order item units based" should be default tax calculation strategy for the "Mobile store" channel
```
Comment (pamil) [inline features/channel/managing_channels/selecting_default_tax_calculation_strategy_for_channel.feature]: `And the default tax calculation strategy for the "Mobile store" channel should be "Order item units based"` seems more natural to me

File: src/Sylius/Behat/Context/Ui/Admin/ManagingChannelsContext.php [hunk 3]
```diff
@@ -442,7 +462,33 @@ public function thePaymentMethodShouldBeAvailableForTheChannel($paymentMethodNam
...
+
+    /**
+     * @Then the default tax zone for the :channel channel should be :taxZone
+     */
+    public function theDefaultTaxZoneForTheChannelShouldBe(ChannelInterface $channel, $taxZone)
+    {
+        $this->updatePage->open(['id' => $channel->getId()]);
+
+        Assert::true(
+            $this->updatePage->isDefaultTaxZoneChosen($taxZone),
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingChannelsContext.php]: `isDefault...` -> `hasDefault...`

File: src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php [hunk 4]
```diff
@@ -20,23 +20,10 @@
 class DefaultTaxZoneProvider implements ZoneProviderInterface
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php]: Default -> ChannelBased?

File: src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php [hunk 5]
```diff
@@ -86,18 +86,20 @@ public function process(OrderInterface $order)
...
      * @return ZoneInterface|null
      */
-    private function getTaxZone(AddressInterface $shippingAddress = null)
+    private function getTaxZone(OrderInterface $order)
     {
+        $shippingAddress = $order->getShippingAddress();
         $zone = null;
+
         if (null !== $shippingAddress) {
             $zone = $this->zoneMatcher->match($shippingAddress);
```
Comment (pamil) [inline src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php]: Why don't we wrap this zone matcher with `TaxZoneProviderInterface`? Then we can setup a composite one, that checks this one first and fallback to channel based tax zone provider, so this method would not be needed (we'll just use `$taxZoneProvider->getZone($order)` to get tax zone for given order).

File: src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php [hunk 6]
```diff
@@ -68,7 +68,7 @@ public function process(OrderInterface $order)
             return;
         }
 
-        $zone = $this->getTaxZone($order->getShippingAddress());
+        $zone = $this->getTaxZone($order);
 
         if (null === $zone) {
             return;
```
Comment (pamil) [inline src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php]: Is it possible not to have a tax zone for an order now? If it's not possible, we should throw an exception.

File: features/channel/managing_channels/selecting_default_tax_zone_for_channel.feature [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    Background:
+        Given there is a zone "EU" containing all members of the European Union
+        And there is a zone "The Rest of the World" containing all other countries
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Adding a new channel with default tax zone
+        Given I want to create a new channel
+        When I specify its code as MOBILE
```
Comment (pamil) [inline features/channel/managing_channels/selecting_default_tax_zone_for_channel.feature]: "MOBILE"

File: src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php (hunk 4 above)
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php]: I think we could leave it as it is for now, as it used as **default** tax zone provider and it could be easily replaced by some other implementation, not necessarily based on channel.

File: src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php (hunk 5 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php]: Totally agreed 👍 However this PR is only about moving *default tax zone* and *default tax calculation strategy* to channel, so such a refactoring could and should be done in separate PR ;)

File: src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php (hunk 6 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Taxation/Processor/OrderTaxesProcessor.php]: I think it still is, as channel could not have default tax zone configured.

File: src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php (hunk 4 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php]: You got my point, we use this class in another service (hmm, there's IoC so we don't really even know if it's this one used) *as* default tax zone provider, but this implementation is based on order's channel state. We can use any other logic in `DefaultTaxZoneProvider` like returning hardcoded tax zone, getting it by address etc., so the name is not even meaningful.

`Default` is one of the keywords that are too broad and unspecific like `Manager`and so on.

Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Provider/DefaultTaxZoneProvider.php]: I think it could be: ``ChannelBasedDefaultTaxZoneProvider`` because the service is not just "zone provider", but it gives you the default zone - when Sylius can't figure it out from the order. ;)

File: src/Sylius/Bundle/CoreBundle/Fixture/Factory/ChannelExampleFactory.php [hunk 8]
```diff
@@ -102,6 +104,7 @@ public function create(array $options = [])
         $channel->setHostname($options['hostname']);
         $channel->setEnabled($options['enabled']);
         $channel->setColor($options['color']);
+        $channel->setDefaultTaxCalculationStrategy($options['default_tax_calculation_strategy']);
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/ChannelExampleFactory.php]: It is not really "Default tax calculation strategy" it is just tax calculation strategy. ;)

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 9]
```diff
@@ -106,6 +106,7 @@ public function theStoreOperatesOnAChannelNamed($channelIdentifier)
     {
         $channel = $this->channelFactory->createNamed($channelIdentifier);
         $channel->setCode($channelIdentifier);
+        $channel->setTaxCalculationStrategy('Order items based');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: Shouldn't be `order_items_based`?

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml [hunk 10]
```diff
@@ -18,13 +18,17 @@
 
     <mapped-superclass name="Sylius\Component\Core\Model\Channel" table="sylius_channel">
         <field name="themeName" column="theme_name" type="string" nullable="true" />
+        <field name="taxCalculationStrategy" type="string" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml]: Missing `column="tax_calculation_strategy"`

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Settings/taxation.html.twig [hunk 11]
```diff
@@ -17,7 +17,6 @@
 {{ form_errors(form) }}
 
 <form action="{{ path('sylius_backend_taxation_settings') }}" method="post" class="form-horizontal" novalidate>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Settings/taxation.html.twig]: The whole template can be removed.

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Settings/taxation.html.twig]: Nevermind it's WebBundle, so it will be removed soon anyway.

File: src/Sylius/Component/Core/Model/Channel.php [hunk 12]
```diff
@@ -14,7 +14,9 @@
 use Doctrine\Common\Collections\ArrayCollection;
 use Doctrine\Common\Collections\Collection;
 use Sylius\Bundle\ThemeBundle\Model\ThemeInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
 use Sylius\Component\Channel\Model\Channel as BaseChannel;
+use Sylius\Component\Core\Taxation\Strategy\TaxCalculationStrategyInterface;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Channel.php]: Redundant.

File: src/Sylius/Component/Core/Model/ChannelInterface.php [hunk 13]
```diff
@@ -68,4 +70,24 @@ public function setDefaultCurrency(CurrencyInterface $currency);
...
+     */
+    public function getDefaultTaxZone();
+
+    /**
+     * @param ZoneInterface $defaultTaxZone
+     */
+    public function setDefaultTaxZone(ZoneInterface $defaultTaxZone);
+
+    /**
+     * @return TaxCalculationStrategyInterface
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/ChannelInterface.php]: Should be `string`

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! Nice work! 👍
